doc/programming/guidelines.rst

```diff
@@ -0,0 +1,291 @@
+similar functionality. This overhead is due to increased lifetimes of
+the units, data and their associated hooks; as well as less potential
+for compiler optimization due to additional internal machinery around
+the recursive calls.
+
+.. rubric::  Inline small nested units
+
+The Spicy compiler is not smart enough yet to inline nested units.
+Since declaration of units incurs additional cost to maintain their
+associated state and hooks, it is advisable to manually inline small
+units where performance is critical.
+
+.. rubric::  Minimize event generation
+
+With Zeek, minimize the number of generated events. Each instance of
+an event comes with overhead as the parsed data needs to be converted
+from Spicy's data model into Zeek's data model, which can involve heap
+allocations even for simple types. It's the number of event instances
+generated at runtime here that matters, *not* the number of event
+types defined in the EVT files (although the latter may increase
+compilation times).
+
+.. rubric::  Aggregate data to be forwarded into other analyzers
+
+When passing chunks of data back into Zeek through
+``zeek::protocol_data_in``, it can be more efficient to aggregate
+multiple chunks inside a temporary variable first, instead of
+forwarding each chunk individually. This is because each chunk
+forwarded to Zeek will go through its analyzer pipeline individually,
+which incurs additional overhead.
+
+.. rubric::  Move fixed local values into global constants
+
+Inside functions and hooks, local variables are created and destroyed
+every time the corresponding code executes. For non-trivial types,
+that can lead to noticeable overhead. If the locals aren't modified,
+consider moving them to global constants instead. (For some particular
+expensive, non-mutable types, Spicy performs this optimization
+internally already; for example, for regular expressions.)
+
+.. rubric::  Avoid small sized byte fields
+
+Avoid using the ``bytes`` type for fields that could be handled using
+integer types. As bytes will always allocate a C++ string under the
+hood, using integer types can improve performance.
+
+.. rubric::  Consider outsourcing into C++
+
+Consider outsourcing complex calculations required for your parsing
+into custom C++ code. In particular, decoding bytes into special
+string representations or peculiar time conversions might be
+significantly faster when implemented in C++ directly. See
+:ref:`extending` for more on how to do that.
+
+.. rubric::  State-management in analyzers
+
+Using global maps to retain analyzer state can cause significant
+memory bloating over long(er) periods. The preferred way is to define
+a :ref:`%context <unit_context>` in top-level units and propagate that
+down to populate any maps or other state inside that context. When
+used with Zeek, Spicy ties the context state to individual connections
+that get teared down when the connection state is removed, thereby
+preventing any state space explosion.
```

Comment:
Even with a context a long-running connection could still lead to unbounded state growth. Maybe rephrase as

```suggestion
Using global maps to retain analyzer state can cause significant
memory bloating over long(er) periods if not managed correctly. If statekeeping cannot be avoided, the preferred way is to define
a :ref:`%context <unit_context>` in top-level units and propagate that
down to populate any maps or other state inside that context. When
used with Zeek, Spicy ties the context state to individual connections
that get teared down when the connection state is removed, thereby
preventing accidental state space explosion. Even when state tables are stored
in a ``%context`` one still needs to manually make sure that its size does not
grow unbounded (e.g., for long-running connections).
```